### ISSUE TITLE
[FEATURE] LoRA Adapter Framework for ThemisDB

### ISSUE THREAD

**Issue by makr-code** [makr-code/ThemisDB#319]
# LoRA Adapter Framework for ThemisDB

## 📋 Summary

Implement a comprehensive LoRA (Low-Rank Adaptation) framework for ThemisDB that enables on-the-fly fine-tuning of LLMs for domain-specific tasks. This issue covers the **general LoRA infrastructure** with `themis_help_lora` as the first application example for documentation assistance.

### Primary Goal
Establish a unified LoRA framework that:
1. **Consolidates general LoRA handling** across ThemisDB
2. **Provides reusable infrastructure** for all LoRA use cases
3. **Implements themis_help_lora** as the first practical application
4. **Enables continuous learning** from user feedback

### Secondary Goals
- Zero-downtime adapter switching
- Multi-tenant adapter support
- Automatic adapter versioning
- Performance monitoring and metrics

---

## 🏗️ Architecture Overview

```
┌─────────────────────────────────────────────────────────────┐
│                    LoRA Framework Core                       │
├─────────────────────────────────────────────────────────────┤
│  ┌──────────────┐  ┌──────────────┐  ┌──────────────┐     │
│  │ Adapter Mgmt │  │ Training Svc │  │ Storage Svc  │     │
│  │              │  │              │  │              │     │
│  │ - Load       │  │ - On-the-fly │  │ - Save       │     │
│  │ - Unload     │  │ - Batch      │  │ - Load       │     │
│  │ - Switch     │  │ - Validate   │  │ - Version    │     │
│  │ - Cache      │  │ - Optimize   │  │ - Metadata   │     │
│  └──────────────┘  └──────────────┘  └──────────────┘     │
└─────────────────────────────────────────────────────────────┘
                              │
                              ▼
┌─────────────────────────────────────────────────────────────┐
│                   Application Layer                          │
├─────────────────────────────────────────────────────────────┤
│  ┌────────────────────────────────────────────────────┐    │
│  │  themis_help_lora (First Use Case)                 │    │
│  │  - Documentation Q&A fine-tuning                   │    │
│  │  - Intent classification improvement               │    │
│  │  - ThemisDB-specific terminology                   │    │
│  │  - User feedback integration                       │    │
│  └────────────────────────────────────────────────────┘    │
│                                                              │
│  ┌────────────────────────────────────────────────────┐    │
│  │  Future Use Cases                                   │    │
│  │  - themis_sql_lora (SQL generation)                │    │
│  │  - themis_ops_lora (Operations assistance)         │    │
│  │  - themis_perf_lora (Performance tuning)           │    │
│  └────────────────────────────────────────────────────┘    │
└─────────────────────────────────────────────────────────────┘
```

---

#### 1.1 Adapter Manager (`lora_adapter_manager.h/cpp`)
```cpp
class LoRAAdapterManager {
public:
    // Lifecycle management
    bool loadAdapter(const std::string& adapter_id);
    bool unloadAdapter(const std::string& adapter_id);
    bool switchAdapter(const std::string& from_id, const std::string& to_id);
    
    // Query and status
    std::vector<std::string> listAdapters();
    AdapterInfo getAdapterInfo(const std::string& adapter_id);
    bool isLoaded(const std::string& adapter_id);
    
    // Performance
    void enableAdapterCache(bool enable);
    CacheStats getCacheStats();
};
```

**Features:**
- Hot-swapping without service restart
- LRU cache for frequently used adapters
- Lazy loading on first use
- Automatic unloading of unused adapters
- Thread-safe operations

#### 1.2 Training Service (`lora_training_service.h/cpp`)
```cpp
class LoRATrainingService {
public:
    // Training modes
    TrainingResult trainOnTheFly(const TrainingData& data);
    TrainingResult trainBatch(const std::vector<TrainingData>& dataset);
    
    // Configuration
    void setTrainingConfig(const LoRAConfig& config);
    LoRAConfig getTrainingConfig();
    
    // Monitoring
    TrainingMetrics getMetrics();
    void registerCallback(TrainingCallback callback);
};
```

#### 1.3 Storage Service (`lora_storage_service.h/cpp`)
```cpp
class LoRAStorageService {
public:
    // Persistence
    bool saveAdapter(const std::string& adapter_id, const AdapterWeights& weights);
    AdapterWeights loadAdapter(const std::string& adapter_id);
    bool deleteAdapter(const std::string& adapter_id);
    
    // Versioning
    std::string createVersion(const std::string& adapter_id);
    bool rollbackToVersion(const std::string& adapter_id, const std::string& version);
    std::vector<std::string> listVersions(const std::string& adapter_id);
    
    // Metadata
    void updateMetadata(const std::string& adapter_id, const json& metadata);
    json getMetadata(const std::string& adapter_id);
};
```

**Storage Options:**
1. **ThemisDB Collection** (Primary): `lora_adapters` collection
2. **File System** (Backup): `data/lora_adapters/{adapter_id}/`
3. **S3/Object Storage** (Optional): For distributed deployments

---

#### 2.1 Use Case: Documentation Assistance Fine-Tuning

**Objectives:**
- Improve accuracy for ThemisDB-specific questions
- Learn from user corrections and feedback
- Adapt to evolving documentation
- Reduce hallucinations on ThemisDB features

#### 2.2 Implementation (`llm/applications/themis_help_lora.h/cpp`)

```cpp
class ThemisHelpLoRA {
public:
    // Query with adapter
    std::string query(const std::string& question);
    
    // Training from feedback
    void addPositiveFeedback(const std::string& question, const std::string& answer);
    void addNegativeFeedback(const std::string& question, const std::string& answer, const std::string& correction);
    
    // Trigger training
    bool trainFromFeedback();
    bool trainFromDocumentation();
    
    // Metrics
    PerformanceMetrics getMetrics();
    FeedbackStats getFeedbackStats();
};
```

#### 2.3 Training Strategy

```
Phase 1: Initial Training (2-4 hours, one-time)
  - Base: Pre-trained LLM
  - Dataset: ThemisDB documentation (1151 docs)
  - Method: Batch training

Phase 2: Continuous Learning (10-30 min, periodic)
  - Trigger: Every 100 feedback items OR daily
  - Dataset: User feedback + corrections
  - Method: Incremental fine-tuning

Phase 3: Version Management
  - Auto-checkpoint after each training
  - A/B testing
  - Rollback if quality degrades
  - Keep last 5 versions
```

---

### Phase 1: Core Framework (2-3 weeks)
- [ ] **Task 1.1**: Implement `LoRAAdapterManager`
  - [ ] Adapter loading/unloading
  - [ ] Hot-swapping mechanism
  - [ ] LRU cache implementation
  - [ ] Thread-safety
  
- [ ] **Task 1.2**: Implement `LoRATrainingService`
  - [ ] On-the-fly training pipeline
  - [ ] Batch training pipeline
  - [ ] Training configuration management
  - [ ] Progress monitoring
  
- [ ] **Task 1.3**: Implement `LoRAStorageService`
  - [ ] ThemisDB collection integration
  - [ ] Versioning system
  - [ ] Metadata management
  - [ ] File system backup
  
- [ ] **Task 1.4**: Configuration and utilities
  - [ ] `LoRAConfig` structure
  - [ ] Validation helpers
  - [ ] Error handling
  - [ ] Logging integration

### Phase 2: themis_help_lora Application (2 weeks)
- [ ] **Task 2.1**: Base implementation
  - [ ] `ThemisHelpLoRA` class
  - [ ] Query interface
  - [ ] Feedback collection
  
- [ ] **Task 2.2**: Training data preparation
  - [ ] Documentation corpus processing
  - [ ] Q&A pair generation
  - [ ] Feedback database schema
  
- [ ] **Task 2.3**: Initial training
  - [ ] Prepare training dataset (1151 docs)
  - [ ] Configure hyperparameters
  - [ ] Run initial training
  - [ ] Validate adapter quality
  
- [ ] **Task 2.4**: Integration with HELP()
  - [ ] Modify `help()` function
  - [ ] Adapter loading logic
  - [ ] Fallback handling
  - [ ] Performance testing

### Phase 3: Continuous Learning (1 week)
- [ ] **Task 3.1**: Feedback collection
  - [ ] Add feedback buttons to responses
  - [ ] Store feedback in database
  - [ ] Validation and filtering
  
- [ ] **Task 3.2**: Incremental training
  - [ ] Trigger mechanism (count/time-based)
  - [ ] Incremental fine-tuning
  - [ ] Checkpoint management
  
- [ ] **Task 3.3**: Quality assurance
  - [ ] A/B testing framework
  - [ ] Quality metrics
  - [ ] Automatic rollback

### Phase 4: Testing & Documentation (1 week)
- [ ] **Task 4.1**: Unit tests
  - [ ] Adapter manager tests
  - [ ] Training service tests
  - [ ] Storage service tests
  - [ ] Integration tests
  
- [ ] **Task 4.2**: Performance tests
  - [ ] Latency benchmarks
  - [ ] Memory usage profiling
  - [ ] Cache effectiveness
  
- [ ] **Task 4.3**: Documentation
  - [ ] API documentation
  - [ ] User guide for themis_help_lora
  - [ ] Developer guide for new adapters
  - [ ] Training best practices

---

### Collection: `lora_adapters`
```json
{
  "_key": "themis_help_lora_v1",
  "adapter_id": "themis_help_lora",
  "version": "1.0.0",
  "base_model": "llama-2-7b",
  "rank": 8,
  "alpha": 16.0,
  "weights_blob": "<binary>",
  "metadata": {
    "description": "Documentation assistance adapter",
    "training_samples": 5000,
    "validation_accuracy": 0.92
  },
  "created_at": "2026-01-11T12:00:00Z",
  "status": "active"
}
```

### Collection: `help_feedback`
```json
{
  "_key": "feedback_12345",
  "question": "How do I enable sharding?",
  "answer": "To enable sharding...",
  "feedback_type": "positive",
  "correction": null,
  "timestamp": "2026-01-11T12:30:00Z",
  "used_for_training": false
}
```

---

### Technical Metrics
- **Adapter Load Time**: < 500ms
- **Training Time**: < 30 min for incremental updates
- **Memory Overhead**: < 200 MB per adapter
- **Inference Latency**: < 100ms increase
- **Cache Hit Rate**: > 80%

### Quality Metrics
- **Accuracy Improvement**: +15-25% over base model
- **Hallucination Reduction**: -50% on ThemisDB topics
- **User Satisfaction**: > 85% positive feedback
- **Correction Rate**: < 10%

---

### Additional Use Cases
1. **themis_sql_lora** - SQL query generation
2. **themis_ops_lora** - Operations troubleshooting
3. **themis_perf_lora** - Performance analysis
4. **themis_security_lora** - Security guidance

### Advanced Features
1. **Multi-tenant Adapters** - Per-customer fine-tuning
2. **Federated Learning** - Distributed training
3. **AutoML for LoRA** - Automatic hyperparameter tuning
4. **Adapter Composition** - Combine multiple adapters

---

## 📚 Resources

- LoRA Paper: https://arxiv.org/abs/2106.09685
- PEFT Library: https://github.com/huggingface/peft
- SafeTensors: https://github.com/huggingface/safetensors

### Related ThemisDB Features
- LLM Integration: `include/llm/embedded_llm.h`
- Documentation Database: `TODO_DOCS_DATABASE_BUILD.md`
- HELP() Function: Current PR (3-tier intent detection)

---

## 📅 Timeline

**Total Duration**: 6-7 weeks

- **Week 1-3**: Core framework implementation
- **Week 4-5**: themis_help_lora application
- **Week 6**: Continuous learning and integration
- **Week 7**: Testing, documentation, deployment

---

## ✅ Definition of Done

- [ ] Core framework fully implemented and tested
- [ ] themis_help_lora successfully trained and deployed
- [ ] Integration with HELP() function complete
- [ ] Continuous learning pipeline operational
- [ ] All tests passing
- [ ] Documentation complete
- [ ] Performance metrics meet targets
- [ ] Code review approved

---

## 💬 Additional Context

Add any other context, screenshots, or examples about the feature request here.

**Comment by makr-code** [comment]
sub issues created
